### ISSUE TITLE
Spurious(?) fallback trigger when using breslow estimator, no issue when running without fallback

### ISSUE THREAD

**Issue by jemus42** [mlr-org/mlr3proba#347]
I'm not entirely sure I'm characterizing this issue correctly, but the gist is:

Using XGBoost with the breslow estimator in `distrcompositor` _and_ a fallback learner results in the fallback being triggered.
Always. Every task I try, every HPC I try (random search with hundreds of evals).

I'm only seeing 

```
Cannot predict, Learner 'surv.xgboost' has not been trained yet\nThis happened PipeOp surv.xgboost's $predict()
```

 yet I can not figure out *why*.

Here's what I've found so far:

| learner | estimator | fallback | encapsulation | error? |
|--------|----------|-----------|--------|---------|
| xgboost | kaplan  | no     | -         | no |
| xgboost | kaplan  | yes   | "callr" | no |
| xgboost | breslow | no   | -          | no |
| xgboost | breslow | yes  | "try"   | no |
| xgboost | breslow | yes  | "callr" | yes |

And here's a reprex comparing some of these versions:

``` r
library("mlr3")
library("mlr3proba")
#devtools::load_all(".")
library("mlr3learners")
library("mlr3pipelines")
requireNamespace("mlr3extralearners")

bl = function(key, ..., .estimator = "kaplan", .fallback = TRUE, .encapsulate = "callr") {
  checkmate::assert_choice(.estimator, choices = c("kaplan", "breslow"))

  learner = lrn(key, ...)
  #fallback = ppl("crankcompositor", lrn("surv.kaplan"), response = TRUE,
  #               method = "mean", overwrite = FALSE, graph_learner = TRUE)
  fallback = lrn("surv.kaplan")

  # Needs to be consistent with each other but doesn't "do" anything, just formality in surv context
  fallback$predict_type = "crank"
  learner$predict_type = "crank"

  graph_learner = po("encode", method = "treatment") %>>%
    po("removeconstants") %>>%
    ppl("distrcompositor", learner = learner, form = "ph", estimator = .estimator, overwrite = FALSE) |>
    as_learner()

  graph_learner$predict_type = "crank"

  if (.fallback) {
    graph_learner$fallback = fallback
    graph_learner$encapsulate = c(train = .encapsulate, predict = .encapsulate)
  }

  graph_learner
}

learners = list(
  XGBCox_fallback_breslow_callr = bl(
    "surv.xgboost", tree_method = "hist", booster = "gbtree",
    objective = "survival:cox",
    max_depth = 2, eta = 0.3, nrounds = 20,
    .estimator = "breslow",
    .fallback = TRUE, .encapsulate = "callr"
  )

  ,

  XGBCox_fallback_breslow_try = bl(
    "surv.xgboost", tree_method = "hist", booster = "gbtree",
    objective = "survival:cox",
    max_depth = 2, eta = 0.3, nrounds = 20,
    .estimator = "breslow",
    .fallback = TRUE, .encapsulate = "try"
  )

  ,

  XGBCox_fallback_kaplan =  bl(
    "surv.xgboost", tree_method = "hist", booster = "gbtree",
    objective = "survival:cox",
    max_depth = 2, eta = 0.3, nrounds = 20,
    .estimator = "kaplan",
    .fallback = TRUE
  )

  ,

  XGBCox_nofallback_breslow = bl(
    "surv.xgboost", tree_method = "hist", booster = "gbtree",
    objective = "survival:cox",
    max_depth = 2, eta = 0.3, nrounds = 20,
    .estimator = "breslow",
    .fallback = FALSE
  )
)

mlr3misc::imap(learners, function(l, id) l$id = id)
#> $XGBCox_fallback_breslow_callr
#> [1] "XGBCox_fallback_breslow_callr"
#> 
#> $XGBCox_fallback_breslow_try
#> [1] "XGBCox_fallback_breslow_try"
#> 
#> $XGBCox_fallback_kaplan
#> [1] "XGBCox_fallback_kaplan"
#> 
#> $XGBCox_nofallback_breslow
#> [1] "XGBCox_nofallback_breslow"

tasks = tsk("rats")

measures = list(msr("surv.cindex", id = "harrell_c"),
  msr("surv.graf", id = "graf_proper_erv", proper = TRUE, ERV = TRUE))


grid = benchmark_grid(
  tasks = tasks,
  learners = learners,
  resamplings = rsmp("holdout", ratio = 4/5)
)

future::plan("multisession", workers = 4)
bmr = benchmark(grid, store_models = TRUE, store_backends = TRUE)
#> INFO  [15:47:50.462] [mlr3] Running benchmark with 4 resampling iterations
#> INFO  [15:47:51.475] [mlr3] Applying learner 'XGBCox_fallback_breslow_callr' on task 'rats' (iter 1/1)
#> INFO  [15:47:52.468] [mlr3] Applying learner 'XGBCox_fallback_breslow_try' on task 'rats' (iter 1/1)
#> INFO  [15:47:53.424] [mlr3] Applying learner 'XGBCox_fallback_kaplan' on task 'rats' (iter 1/1)
#> INFO  [15:47:54.402] [mlr3] Applying learner 'XGBCox_nofallback_breslow' on task 'rats' (iter 1/1)
#> INFO  [15:47:56.540] [mlr3] Finished benchmark
```

```r
# errors column indicates there's something wrong
bmr
#> <BenchmarkResult> of 4 rows with 4 resampling runs
#>  nr task_id                    learner_id resampling_id iters warnings errors
#>   1    rats XGBCox_fallback_breslow_callr       holdout     1        0      1
#>   2    rats   XGBCox_fallback_breslow_try       holdout     1        0      0
#>   3    rats        XGBCox_fallback_kaplan       holdout     1        0      0
#>   4    rats     XGBCox_nofallback_breslow       holdout     1        0      0

aggr = bmr$aggregate(measures = measures, conditions = TRUE)

# KM = 0.5 and graf_proper with ERV = 0 imply that fallback was triggered
aggr
#>    nr task_id                    learner_id resampling_id iters warnings errors
#> 1:  1    rats XGBCox_fallback_breslow_callr       holdout     1        0      1
#> 2:  2    rats   XGBCox_fallback_breslow_try       holdout     1        0      0
#> 3:  3    rats        XGBCox_fallback_kaplan       holdout     1        0      0
#> 4:  4    rats     XGBCox_nofallback_breslow       holdout     1        0      0
#>    harrell_c graf_proper_erv
#> 1: 0.5000000        0.000000
#> 2: 0.8078358       -2.763054
#> 3: 0.8078358       -1.134256
#> 4: 0.8078358       -2.763054
#> Hidden columns: resample_result
````

```r
# Error in question I can't figure out -----------------------------------

learners$XGBCox_fallback_breslow_callr$train(tasks)
pred = learners$XGBCox_fallback_breslow_callr$predict(tasks)
pred$score(measures, task = tasks, train_set = tasks$row_ids)
#>       harrell_c graf_proper_erv 
#>             0.5             0.0
learners$XGBCox_fallback_breslow_callr$state$log
#>      stage  class
#> 1: predict  error
#> 2: predict output
#>                                                                                                                msg
#> 1: Cannot predict, Learner 'surv.xgboost' has not been trained yet\nThis happened PipeOp surv.xgboost's $predict()
#> 2:                                                                          Using fallback learner for predictions

# Doesn't appear without fallback
learners$XGBCox_nofallback_breslow$train(tasks)
learners$XGBCox_nofallback_breslow$predict(tasks)
#> <PredictionSurv> for 300 observations:
#>     row_ids time status       crank          lp     distr
#>           1  101  FALSE -0.04284317 -0.04284317 <list[1]>
#>           2   49   TRUE -0.74990779 -0.74990779 <list[1]>
#>           3  104  FALSE -0.74990779 -0.74990779 <list[1]>
#> ---                                                      
#>         298   92  FALSE -0.99841115 -0.99841115 <list[1]>
#>         299  104  FALSE -0.99841115 -0.99841115 <list[1]>
#>         300  102  FALSE -0.99841115 -0.99841115 <list[1]>
learners$XGBCox_nofallback_breslow$state$log
#> Empty data.table (0 rows and 3 cols): stage,class,msg

# Nor with fallback but without breslow
learners$XGBCox_fallback_kaplan$train(tasks)
learners$XGBCox_fallback_kaplan$predict(tasks)
#> <PredictionSurv> for 300 observations:
#>     row_ids time status       crank          lp     distr
#>           1  101  FALSE -0.04284317 -0.04284317 <list[1]>
#>           2   49   TRUE -0.74990779 -0.74990779 <list[1]>
#>           3  104  FALSE -0.74990779 -0.74990779 <list[1]>
#> ---                                                      
#>         298   92  FALSE -0.99841115 -0.99841115 <list[1]>
#>         299  104  FALSE -0.99841115 -0.99841115 <list[1]>
#>         300  102  FALSE -0.99841115 -0.99841115 <list[1]>
learners$XGBCox_fallback_kaplan$state$log
#> Empty data.table (0 rows and 3 cols): stage,class,msg
```

<sup>Created on 2024-01-18 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.3.2 (2023-10-31)
#>  os       macOS Sonoma 14.2.1
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Berlin
#>  date     2024-01-18
#>  pandoc   3.1.1 @ /System/Volumes/Data/Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package           * version     date (UTC) lib source
#>  backports           1.4.1       2021-12-13 [1] CRAN (R 4.3.0)
#>  cachem              1.0.8       2023-05-01 [1] CRAN (R 4.3.0)
#>  callr               3.7.3       2022-11-02 [1] CRAN (R 4.3.0)
#>  checkmate           2.3.1       2023-12-04 [1] CRAN (R 4.3.1)
#>  cli                 3.6.2       2023-12-11 [1] CRAN (R 4.3.1)
#>  codetools           0.2-19      2023-02-01 [2] CRAN (R 4.3.2)
#>  colorspace          2.1-0       2023-01-23 [1] CRAN (R 4.3.0)
#>  crayon              1.5.2       2022-09-29 [1] CRAN (R 4.3.0)
#>  data.table          1.14.10     2023-12-08 [1] CRAN (R 4.3.1)
#>  devtools            2.4.5       2022-10-11 [1] CRAN (R 4.3.0)
#>  dictionar6          0.1.3       2021-09-13 [1] CRAN (R 4.3.0)
#>  digest              0.6.33      2023-07-07 [1] CRAN (R 4.3.0)
#>  distr6              1.8.4       2023-11-13 [1] Github (xoopR/distr6@1854b22)
#>  dplyr               1.1.4       2023-11-17 [1] CRAN (R 4.3.1)
#>  ellipsis            0.3.2       2021-04-29 [1] CRAN (R 4.3.0)
#>  evaluate            0.23        2023-11-01 [1] CRAN (R 4.3.1)
#>  fansi               1.0.6       2023-12-08 [1] CRAN (R 4.3.1)
#>  fastmap             1.1.1       2023-02-24 [1] CRAN (R 4.3.0)
#>  fs                  1.6.3       2023-07-20 [1] CRAN (R 4.3.0)
#>  future              1.33.1      2023-12-22 [1] CRAN (R 4.3.1)
#>  future.apply        1.11.1      2023-12-21 [1] CRAN (R 4.3.1)
#>  generics            0.1.3       2022-07-05 [1] CRAN (R 4.3.0)
#>  ggplot2             3.4.4       2023-10-12 [1] CRAN (R 4.3.1)
#>  globals             0.16.2      2022-11-21 [1] CRAN (R 4.3.0)
#>  glue                1.7.0       2024-01-09 [1] CRAN (R 4.3.1)
#>  gtable              0.3.4       2023-08-21 [1] CRAN (R 4.3.0)
#>  htmltools           0.5.7       2023-11-03 [1] CRAN (R 4.3.1)
#>  htmlwidgets         1.6.4       2023-12-06 [1] CRAN (R 4.3.1)
#>  httpuv              1.6.13      2023-12-06 [1] CRAN (R 4.3.1)
#>  jsonlite            1.8.8       2023-12-04 [1] CRAN (R 4.3.1)
#>  knitr               1.45        2023-10-30 [1] CRAN (R 4.3.1)
#>  later               1.3.2       2023-12-06 [1] CRAN (R 4.3.1)
#>  lattice             0.21-9      2023-10-01 [1] CRAN (R 4.3.1)
#>  lgr                 0.4.4       2022-09-05 [1] CRAN (R 4.3.0)
#>  lifecycle           1.0.4       2023-11-07 [1] CRAN (R 4.3.1)
#>  listenv             0.9.0       2022-12-16 [1] CRAN (R 4.3.0)
#>  magrittr            2.0.3       2022-03-30 [1] CRAN (R 4.3.0)
#>  Matrix              1.6-1.1     2023-09-18 [1] CRAN (R 4.3.1)
#>  memoise             2.0.1       2021-11-26 [1] CRAN (R 4.3.0)
#>  mime                0.12        2021-09-28 [1] CRAN (R 4.3.0)
#>  miniUI              0.1.1.1     2018-05-18 [1] CRAN (R 4.3.0)
#>  mlr3              * 0.17.2      2024-01-09 [1] CRAN (R 4.3.1)
#>  mlr3extralearners   0.7.1-9000  2024-01-11 [1] Github (mlr-org/mlr3extralearners@0cbdb72)
#>  mlr3learners      * 0.5.8       2023-12-21 [1] CRAN (R 4.3.1)
#>  mlr3misc            0.13.0-9000 2023-10-17 [1] Github (mlr-org/mlr3misc@43c0ffa)
#>  mlr3pipelines     * 0.5.0-2     2023-12-08 [1] CRAN (R 4.3.1)
#>  mlr3proba           0.5.8       2024-01-16 [1] Github (mlr-org/mlr3proba@8565481)
#>  mlr3viz             0.7.0.9000  2023-12-22 [1] Github (mlr-org/mlr3viz@73aa4ec)
#>  munsell             0.5.0       2018-06-12 [1] CRAN (R 4.3.0)
#>  ooplah              0.2.0       2022-01-21 [1] CRAN (R 4.3.0)
#>  palmerpenguins      0.1.1       2022-08-15 [1] CRAN (R 4.3.0)
#>  paradox             0.11.1      2023-03-17 [1] CRAN (R 4.3.0)
#>  parallelly          1.36.0      2023-05-26 [1] CRAN (R 4.3.0)
#>  param6              0.2.4       2023-07-24 [1] Github (xoopR/param6@0fa3577)
#>  pillar              1.9.0       2023-03-22 [1] CRAN (R 4.3.0)
#>  pkgbuild            1.4.3       2023-12-10 [1] CRAN (R 4.3.1)
#>  pkgconfig           2.0.3       2019-09-22 [1] CRAN (R 4.3.0)
#>  pkgload             1.3.3       2023-09-22 [1] CRAN (R 4.3.1)
#>  processx            3.8.3       2023-12-10 [1] CRAN (R 4.3.1)
#>  profvis             0.3.8       2023-05-02 [1] CRAN (R 4.3.0)
#>  promises            1.2.1       2023-08-10 [1] CRAN (R 4.3.0)
#>  ps                  1.7.5       2023-04-18 [1] CRAN (R 4.3.0)
#>  purrr               1.0.2       2023-08-10 [1] CRAN (R 4.3.0)
#>  R.cache             0.16.0      2022-07-21 [1] CRAN (R 4.3.0)
#>  R.methodsS3         1.8.2       2022-06-13 [1] CRAN (R 4.3.0)
#>  R.oo                1.25.0      2022-06-12 [1] CRAN (R 4.3.0)
#>  R.utils             2.12.3      2023-11-18 [1] CRAN (R 4.3.1)
#>  R6                  2.5.1       2021-08-19 [1] CRAN (R 4.3.0)
#>  Rcpp                1.0.12      2024-01-09 [1] CRAN (R 4.3.1)
#>  remotes             2.4.2.1     2023-07-18 [1] CRAN (R 4.3.0)
#>  reprex              2.1.0       2024-01-11 [1] CRAN (R 4.3.1)
#>  RhpcBLASctl         0.23-42     2023-02-11 [1] CRAN (R 4.3.0)
#>  rlang               1.1.3       2024-01-10 [1] CRAN (R 4.3.1)
#>  rmarkdown           2.25        2023-09-18 [1] CRAN (R 4.3.1)
#>  rprojroot           2.0.4       2023-11-05 [1] CRAN (R 4.3.1)
#>  rstudioapi          0.15.0      2023-07-07 [1] CRAN (R 4.3.0)
#>  scales              1.3.0       2023-11-28 [1] CRAN (R 4.3.1)
#>  sessioninfo         1.2.2       2021-12-06 [1] CRAN (R 4.3.0)
#>  set6                0.2.6       2023-10-16 [1] Github (xoopR/set6@a901255)
#>  shiny               1.8.0       2023-11-17 [1] CRAN (R 4.3.1)
#>  stringi             1.8.3       2023-12-11 [1] CRAN (R 4.3.1)
#>  stringr             1.5.1       2023-11-14 [1] CRAN (R 4.3.1)
#>  styler              1.10.2      2023-12-12 [1] Github (r-lib/styler@e987d7a)
#>  survival            3.5-7       2023-08-14 [1] CRAN (R 4.3.0)
#>  survivalmodels      0.1.18      2023-11-21 [1] Github (RaphaelS1/survivalmodels@79b913f)
#>  tibble              3.2.1       2023-03-20 [1] CRAN (R 4.3.0)
#>  tidyselect          1.2.0       2022-10-10 [1] CRAN (R 4.3.0)
#>  urlchecker          1.0.1       2021-11-30 [1] CRAN (R 4.3.0)
#>  usethis             2.2.2       2023-07-06 [1] CRAN (R 4.3.0)
#>  utf8                1.2.4       2023-10-22 [1] CRAN (R 4.3.1)
#>  uuid                1.1-1       2023-08-17 [1] CRAN (R 4.3.0)
#>  vctrs               0.6.5       2023-12-01 [1] CRAN (R 4.3.1)
#>  withr               2.5.2       2023-10-30 [1] CRAN (R 4.3.1)
#>  xfun                0.41        2023-11-01 [1] CRAN (R 4.3.1)
#>  xgboost             1.7.6.1     2023-12-06 [1] CRAN (R 4.3.1)
#>  xtable              1.8-4       2019-04-21 [1] CRAN (R 4.3.0)
#>  yaml                2.3.8       2023-12-11 [1] CRAN (R 4.3.1)
#> 
#>  [1] /Users/Lukas/Library/R/arm64/4.3/library
#>  [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

**Comment by jemus42** [comment]
Adressed by #349 🙏🏻 🎉 

While I appreciate the solution, I'd be curious to understand the problem.

I mean, why does the issue only pop up when `callr` is used, but not when `try` is used for encapsulation?

**Comment by jemus42** [comment]
Fixed by #349
